server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -57,12 +60,52 @@
             Response extends ReplicationResponse & WriteResponse
         > extends TransportReplicationAction<Request, ReplicaRequest, Response> {
 
+    private final boolean forceExecutionOnPrimary;
+    private final WriteMemoryLimits writeMemoryLimits;
+    private final String executor;
+
     protected TransportWriteAction(Settings settings, String actionName, TransportService transportService,
                                    ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool,
                                    ShardStateAction shardStateAction, ActionFilters actionFilters, Writeable.Reader<Request> request,
-                                   Writeable.Reader<ReplicaRequest> replicaRequest, String executor, boolean forceExecutionOnPrimary) {
+                                   Writeable.Reader<ReplicaRequest> replicaRequest, String executor, boolean forceExecutionOnPrimary,
+                                   WriteMemoryLimits writeMemoryLimits) {
         super(settings, actionName, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
-              request, replicaRequest, executor, true, forceExecutionOnPrimary);
+            request, replicaRequest, ThreadPool.Names.SAME, true, forceExecutionOnPrimary);
+        this.executor = executor;
+        this.forceExecutionOnPrimary = forceExecutionOnPrimary;
+        this.writeMemoryLimits = writeMemoryLimits;
+    }
+
+    @Override
+    protected Releasable checkOperationLimits(Request request) {
+        if (coordinatingBytesNeedAccounted(request)) {
+            long operationSizeInBytes = primaryOperationSize(request);
+            return writeMemoryLimits.markCoordinatingOperationStarted(operationSizeInBytes);
+        } else {
+            return () -> {};
+        }
+    }
+
+    protected boolean coordinatingBytesNeedAccounted(Request request) {
+        return false;
+    }
+
+    @Override
+    protected Releasable checkPrimaryLimits(Request request) {
+        return writeMemoryLimits.markPrimaryOperationStarted(primaryOperationSize(request));
+    }
+
+    protected long primaryOperationSize(Request request) {
+        return 0;
```

Comment:
This will call `writeMemoryLimits.markPrimaryOperationStarted(0)` for those actions where we possibly don't want to do accounting (and breaking) for. I wonder if this will lead to confusion, and we might start rejecting requests later when size = 0 but we are already above whatever memory threshold we define.

Reply:
I thought we want to account some for every op? I understand that we might need special logic to allow things when we implement rejections.

Reply:
ok, let's move this discussion to a follow-up PR where we introduce rejections. I'm mostly worried that we would forget about the callers that are using `0`